src/Dfc.CourseDirectory.Web/Controllers/PublishApprenticeships/PublishApprenticeshipsController.cs

```diff
@@ -19,43 +20,52 @@ namespace Dfc.CourseDirectory.Web.Controllers.PublishApprenticeships
     {
-        private ISession Session => HttpContext.Session;
         private readonly IApprenticeshipService _apprenticeshipService;
+        private readonly ICosmosDbQueryDispatcher _cosmosDbQueryDispatcher;
 
-        public PublishApprenticeshipsController(IApprenticeshipService apprenticeshipService)
-        {
-            if (apprenticeshipService == null)
-            {
-                throw new ArgumentNullException(nameof(apprenticeshipService));
-            }
+        private ISession Session => HttpContext.Session;
 
+        public PublishApprenticeshipsController(
+            IApprenticeshipService apprenticeshipService,
+            ICosmosDbQueryDispatcher cosmosDbQueryDispatcher)
+        {
             _apprenticeshipService = apprenticeshipService;
+            _cosmosDbQueryDispatcher = cosmosDbQueryDispatcher;
         }
 
         [Authorize]
         [HttpGet]
-        public IActionResult Index()
+        public async Task<IActionResult> Index()
         {
-            PublishApprenticeshipsViewModel vm = new PublishApprenticeshipsViewModel();
             int? UKPRN = Session.GetInt32("UKPRN");
+
             if (!UKPRN.HasValue)
+            {
                 return RedirectToAction("Index", "Home", new { errmsg = "Please select a Provider." });
-            var apprenticeships = _apprenticeshipService.GetApprenticeshipByUKPRN(UKPRN.Value.ToString()).Result.Value.Where(x => x.RecordStatus == RecordStatus.BulkUploadPending);
-
-            var getApprenticeships = _apprenticeshipService.GetApprenticeshipByUKPRN(UKPRN.Value.ToString()).Result.Value.Where(x => x.ProviderUKPRN == UKPRN);
+            }
 
-            vm.ListOfApprenticeships = apprenticeships;
-            vm.ListOfApprenticeships = GetErrorMessages(vm.ListOfApprenticeships);
+            PublishApprenticeshipsViewModel vm = new PublishApprenticeshipsViewModel();
 
-            if (!getApprenticeships.Any(x => x.RecordStatus == RecordStatus.BulkUploadPending))
+            var apprenticeships = await _cosmosDbQueryDispatcher.ExecuteQuery(new GetApprenticeships
             {
-                vm.AreAllReadyToBePublished = true;
-            }
+                Predicate = a =>
+                    a.ProviderUKPRN == UKPRN
+                    && (a.RecordStatus == (int)ApprenticeshipStatus.BulkUploadPending || a.RecordStatus == (int)ApprenticeshipStatus.BulkUploadReadyToGoLive)
+            });
+
+            var bulkUploadPendingApprenticeships = apprenticeships.Values.Where(a => a.RecordStatus == (int)ApprenticeshipStatus.BulkUploadPending).Select(Apprenticeship.FromCosmosDbModel).ToArray();
+            var bulkUploadReadyToGoLiveApprenticeships = apprenticeships.Values.Where(a => a.RecordStatus == (int)ApprenticeshipStatus.BulkUploadReadyToGoLive).Select(Apprenticeship.FromCosmosDbModel).ToArray();
 
-            if (vm.AreAllReadyToBePublished)
+            vm.ListOfApprenticeships = GetErrorMessages(bulkUploadPendingApprenticeships);
+
```

Comment:
Does the `AreAllReadyToBePublished` property need to be populated on the VM?

Reply:
Good spot